examples/stable_diffusion_v2/ldm/modules/attention.py

```diff
@@ -109,13 +109,27 @@ def __init__(self, query_dim, context_dim=None, heads=8, dim_head=64, dropout=1.
         self.reshape = ops.Reshape()
         self.softmax = ops.Softmax(axis=-1)
         self.transpose = ops.Transpose()
-        self.to_q = nn.Dense(query_dim, inner_dim, has_bias=False).to_float(dtype)
-        self.to_k = nn.Dense(context_dim, inner_dim, has_bias=False).to_float(dtype)
-        self.to_v = nn.Dense(context_dim, inner_dim, has_bias=False).to_float(dtype)
+        # cong edit here, to add lora matrices
+        # self.to_q = nn.Dense(query_dim, inner_dim, has_bias=False).to_float(dtype)
+        # self.to_k = nn.Dense(context_dim, inner_dim, has_bias=False).to_float(dtype)
+        # self.to_v = nn.Dense(context_dim, inner_dim, has_bias=False).to_float(dtype)
+
+        # self.to_out = nn.SequentialCell(
+        #     nn.Dense(inner_dim, query_dim).to_float(dtype),
+        #     nn.Dropout(dropout) if is_old_ms_version() else nn.Dropout(p=1-dropout)
+        #     )
+
+        # replace Dense Layer with LoRADense
+        self.to_q = LoRADense(in_channels=query_dim, out_channels=inner_dim, lora_rank=8, lora_alpha=16, has_bias=False).to_float(dtype)
+        self.to_k = LoRADense(in_channels=context_dim, out_channels=inner_dim, lora_rank=8, lora_alpha=16, has_bias=False).to_float(dtype)
+        self.to_v = LoRADense(in_channels=context_dim, out_channels=inner_dim, lora_rank=8, lora_alpha=16, has_bias=False).to_float(dtype)
+        # if distributed training is required, invoke shard method  
+
         self.to_out = nn.SequentialCell(
-            nn.Dense(inner_dim, query_dim).to_float(dtype),
+            LoRADense(inner_dim, query_dim, lora_rank=8, lora_alpha=16).to_float(dtype),
```

Comment:
If we want to train UNet or do basic fine-tuning, we don't need the low rank matrices in LoRA. So we should replace the original dense layer with lora layer by choice. 

A better impl should use an injector or wrapper to inject lora layers in the network, instead of hacking into the attention layer.